storage/raftstorage.go

```diff
@@ -67,80 +67,88 @@ func (r *Range) InitialState() (raftpb.HardState, raftpb.ConfState, error) {
 	return hs, cs, nil
 }
 
-// loadLastIndex looks in the engine to find the last log index.
+// setLastIndex persists a new last index and updates its cached
+// value in the range. If a non-nil engine is passed, it is used for the
+// write; otherwise the range's default engine is used. Commit() will
```

Comment:
There's no such thing as "as atomic as possible"; it's either atomic or not. It's ok for r.lastIndex to lag behind the on-disk state. It is important, on the other hand, that the on-disk lastIndex is updated atomically with the writing of that last log entry. This method shouldn't commit anything; just execute an MVCCPut on the supplied engine. It also shouldn't touch `r.lastIndex`; leave that to be done after the caller has committed its batch.